File: android/app/src/main/java/my/ssdid/wallet/feature/invite/InviteAcceptViewModel.kt
package my.ssdid.wallet.feature.invite

import android.net.Uri
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import my.ssdid.wallet.domain.model.Did
import my.ssdid.wallet.domain.model.Identity
import my.ssdid.wallet.domain.transport.SsdidHttpClient
import my.ssdid.wallet.domain.transport.dto.AcceptWithWalletRequest
import my.ssdid.wallet.domain.transport.dto.InvitationDetailsResponse
import my.ssdid.wallet.domain.vault.Vault
import my.ssdid.wallet.domain.verifier.Verifier
import my.ssdid.wallet.platform.security.UrlValidator
import javax.inject.Inject

data class InviteAcceptUiState(
    val isLoading: Boolean = true,
    val invitation: InvitationDetailsResponse? = null,
    val emailMatch: Boolean = false,
    val walletEmail: String = "",
    val error: String? = null,
    val isAccepting: Boolean = false,
    val acceptSuccess: Boolean = false,
    val sessionToken: String? = null,
    val callbackUri: Uri? = null
)

@HiltViewModel
class InviteAcceptViewModel @Inject constructor(
    private val vault: Vault,
    private val httpClient: SsdidHttpClient,
    private val verifier: Verifier,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val json = Json { ignoreUnknownKeys = true }

    val serverUrl: String = savedStateHandle["serverUrl"] ?: ""
    val token: String = savedStateHandle["token"] ?: ""
    val callbackUrl: String = savedStateHandle["callbackUrl"] ?: ""

    private val _state = MutableStateFlow(InviteAcceptUiState())
    val state = _state.asStateFlow()

    private val _selectedIdentity = MutableStateFlow<Identity?>(null)
    val selectedIdentity = _selectedIdentity.asStateFlow()

    init {
        loadInvitation()
    }

    private fun loadInvitation() {
        viewModelScope.launch {
            try {
                if (!UrlValidator.isValidServerUrl(serverUrl)) {
                    _state.update { it.copy(
                        isLoading = false,
                        error = "Invalid server URL"
                    ) }
                    return@launch
                }

                val api = httpClient.serverApi(serverUrl)
                val invitation = api.getInvitationByToken(token)

                val identities = vault.listIdentities()
                val firstIdentity = identities.firstOrNull()
                _selectedIdentity.value = firstIdentity
                val walletEmail = firstIdentity?.email ?: ""

                val emailMatch = walletEmail.isNotBlank() &&
                    walletEmail.equals(invitation.email, ignoreCase = true)

                _state.update { it.copy(
                    isLoading = false,
                    invitation = invitation,
                    emailMatch = emailMatch,
                    walletEmail = walletEmail,
                    error = if (!emailMatch && walletEmail.isNotBlank())
                        "Email mismatch: invitation is for ${invitation.email} but your wallet email is $walletEmail"
                    else if (walletEmail.isBlank())
                        "No email configured in wallet profile"
                    else null
                ) }
            } catch (e: Exception) {
                io.sentry.Sentry.captureException(e)
                val message = when {
                    e is retrofit2.HttpException && e.code() == 404 ->
                        "Invitation not found or expired"
                    else -> e.message ?: "Failed to load invitation"
                }
                _state.update { it.copy(
                    isLoading = false,
                    error = message
                ) }
            }
        }
    }

    fun accept() {
        if (_state.value.isAccepting) return

        viewModelScope.launch {
            _state.update { it.copy(isAccepting = true, error = null) }
            try {
                val selectedDid = _selectedIdentity.value?.did
                    ?: run {
                        _state.update { it.copy(isAccepting = false, error = "No identity selected") }
                        return@launch
                    }

                val credential = vault.getCredentialForDid(selectedDid)
                    ?: run {
                        _state.update { it.copy(
                            isAccepting = false,
                            error = "No credential found for selected identity. Please register with a service first."
                        ) }
                        return@launch
                    }
                val credentialJson = json.encodeToJsonElement(
                    my.ssdid.wallet.domain.model.VerifiableCredential.serializer(),
                    credential
                )

                val api = httpClient.serverApi(serverUrl)
                val response = api.acceptWithWallet(
                    token,
                    AcceptWithWalletRequest(
                        credential = credentialJson,
                        email = _selectedIdentity.value?.email ?: _state.value.walletEmail
                    )
                )

                // Validate server DID format
                Did.validate(response.serverDid).onFailure {
                    _state.update { s -> s.copy(
                        isAccepting = false,
                        error = "Invalid server DID: ${it.message}"
                    ) }
                    return@launch
                }

                // Verify server signature — blank fields = fatal (possible MITM)
                if (response.serverDid.isBlank() || response.serverSignature.isBlank()) {
                    _state.update { it.copy(
                        isAccepting = false,
                        error = "Server did not provide identity proof. Please try again."
                    ) }
                    return@launch
                }

                val verified = verifier.verifyChallengeResponse(
                    response.serverDid,
                    response.serverKeyId,
                    response.sessionToken,
                    response.serverSignature
                ).getOrThrow()

                if (!verified) {
                    _state.update { it.copy(
                        isAccepting = false,
                        error = "Server identity verification failed. Please try again."
                    ) }
                    return@launch
                }

                val callbackUri = if (callbackUrl.isNotBlank()) {
                    Uri.parse("$callbackUrl?session_token=${Uri.encode(response.sessionToken)}&status=success")
                } else null

                _state.update { it.copy(
                    isAccepting = false,
                    acceptSuccess = true,
                    sessionToken = response.sessionToken,
                    callbackUri = callbackUri
                ) }
            } catch (e: Exception) {
                io.sentry.Sentry.captureException(e)
                _state.update { it.copy(
                    isAccepting = false,
                    error = e.message ?: "Failed to accept invitation"
                ) }
            }
        }
    }

    fun selectIdentity(identity: Identity) {
        _selectedIdentity.value = identity
        _state.update { it.copy(
            walletEmail = identity.email ?: "",
            emailMatch = !identity.email.isNullOrBlank() &&
                identity.email.equals(_state.value.invitation?.email, ignoreCase = true),
            error = if (!identity.email.isNullOrBlank() &&
                !identity.email.equals(_state.value.invitation?.email, ignoreCase = true))
                "Email mismatch: invitation is for ${_state.value.invitation?.email} but identity email is ${identity.email}"
            else if (identity.email.isNullOrBlank())
                "No email configured for this identity"
            else null
        ) }
    }

    fun decline() {
        val callbackUri = if (callbackUrl.isNotBlank()) {
            Uri.parse("$callbackUrl?status=cancelled")
        } else null
        _state.update { it.copy(callbackUri = callbackUri) }
    }

    fun buildErrorCallbackUri(message: String): Uri? {
        return if (callbackUrl.isNotBlank()) {
            Uri.parse("$callbackUrl?status=error&message=${Uri.encode(message)}")
        } else null
    }
}

File: ios/SsdidWallet/Feature/Invite/InviteAcceptViewModel.swift
import Foundation
import UIKit

@Observable
@MainActor
final class InviteAcceptViewModel {

    // MARK: - State

    var isLoading = true
    var invitation: InvitationDetailsResponse?
    var identities: [Identity] = []
    var selectedIdentity: Identity?
    var walletEmail = ""
    var emailMatch = false
    var errorMessage: String?
    var isAccepting = false
    var isSuccess = false
    var sessionToken: String?
    var shouldAutoReturn = false

    // MARK: - Private

    private let serverUrl: String
    private let token: String
    private let callbackUrl: String
    private let services: ServiceContainer
    private var acceptTask: Task<Void, Never>?
    private var hasReturned = false

    init(serverUrl: String, token: String, callbackUrl: String, services: ServiceContainer) {
        self.serverUrl = serverUrl
        self.token = token
        self.callbackUrl = callbackUrl
        self.services = services
    }

    // MARK: - Public

    func cancelAccept() {
        acceptTask?.cancel()
    }

    func loadData() async {
        isLoading = true
        errorMessage = nil

        do {
            let driveApi = services.httpClient.driveApi(baseURL: serverUrl)
            let inv = try await driveApi.getInvitationByToken(token)
            invitation = inv

            // Load identities
            identities = await services.vault.listIdentities().filter { $0.isActive }
            if identities.count == 1 {
                selectedIdentity = identities.first
            }

            // Use selected identity's email
            walletEmail = selectedIdentity?.email ?? ""

            let normalizeEmail: (String) -> String = {
                $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            }
            emailMatch = !walletEmail.isEmpty && normalizeEmail(walletEmail) == normalizeEmail(inv.email)

            if !emailMatch && !walletEmail.isEmpty {
                errorMessage = "Email mismatch: invitation is for \(inv.email) but your wallet email is \(walletEmail)"
            } else if walletEmail.isEmpty {
                errorMessage = "No email configured in wallet profile"
            }
        } catch {
            guard !Task.isCancelled else { return }
            if let httpError = error as? HttpError,
               case .requestFailed(let statusCode, _) = httpError, statusCode == 404 {
                errorMessage = "Invitation not found or expired"
            } else {
                errorMessage = error.localizedDescription
            }
        }

        guard !Task.isCancelled else { return }
        isLoading = false
    }

    func accept() {
        guard !isAccepting, let identity = selectedIdentity else { return }

        isAccepting = true
        errorMessage = nil

        acceptTask?.cancel()
        acceptTask = Task {
            defer { isAccepting = false }
            do {
                let driveApi = services.httpClient.driveApi(baseURL: serverUrl)

                // Get or register credential
                var credential = await services.vault.getCredentialForDid(identity.did)
                if credential == nil {
                    credential = try await registerWithDrive(identity: identity, driveApi: driveApi)
                }
                guard let vc = credential else {
                    errorMessage = "Failed to obtain credential"
                    return
                }

                // Accept with wallet
                let response = try await driveApi.acceptWithWallet(
                    token: token,
                    request: AcceptWithWalletRequest(credential: vc, email: walletEmail)
                )

                // Verify server signature -- blank fields = fatal
                guard !response.serverDid.isEmpty, !response.serverSignature.isEmpty else {
                    errorMessage = "Server did not provide identity proof. Please try again."
                    return
                }

                let verified = try await services.ssdidClient.verifyChallengeResponse(
                    did: response.serverDid,
                    keyId: response.serverKeyId,
                    challenge: response.sessionToken,
                    signedChallenge: response.serverSignature
                )

                guard verified else {
                    errorMessage = "Server identity verification failed. Please try again."
                    return
                }

                sessionToken = response.sessionToken
                isSuccess = true

                // Auto-return after short delay
                if !callbackUrl.isEmpty {
                    try await Task.sleep(for: .seconds(1))
                    guard !Task.isCancelled else { return }
                    shouldAutoReturn = true
                }
            } catch {
                if !Task.isCancelled {
                    errorMessage = error.localizedDescription
                }
            }
        }
    }

    func returnToDrive(sessionToken: String, pop: () -> Void) {
        guard !hasReturned else { return }
        hasReturned = true

        if callbackUrl.isEmpty {
            pop()
            return
        }

        guard var components = URLComponents(string: callbackUrl) else {
            pop()
            return
        }

        var items = components.queryItems ?? []
        items.append(URLQueryItem(name: "session_token", value: sessionToken))
        items.append(URLQueryItem(name: "status", value: "success"))
        components.queryItems = items

        if let url = components.url, UIApplication.shared.canOpenURL(url) {
            UIApplication.shared.open(url)
        }
        pop()
    }

    func decline(pop: () -> Void) {
        if !callbackUrl.isEmpty {
            guard var components = URLComponents(string: callbackUrl) else {
                pop()
                return
            }
            var items = components.queryItems ?? []
            items.append(URLQueryItem(name: "status", value: "cancelled"))
            components.queryItems = items
            if let url = components.url, UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.open(url)
            }
        }
        pop()
    }

    // MARK: - Private

    private func registerWithDrive(identity: Identity, driveApi: DriveApi) async throws -> VerifiableCredential {
        let startResp = try await driveApi.register(request: RegisterStartRequest(
            did: identity.did,
            keyId: identity.keyId
        ))
        let signatureBytes = try await services.vault.sign(keyId: identity.keyId, data: Data(startResp.challenge.utf8))
        let signedChallenge = Multibase.encode(signatureBytes)
        let profileClaims = identity.claimsMap()
        let verifyResp = try await driveApi.registerVerify(request: RegisterVerifyRequest(
            did: identity.did,
            keyId: identity.keyId,
            signedChallenge: signedChallenge,
            inviteToken: token,
            sharedClaims: profileClaims.isEmpty ? nil : profileClaims
        ))
        let vc = verifyResp.credential
        try await services.vault.storeCredential(vc)
        return vc
    }
}
